src/main/java/org/broadinstitute/dsm/DSMServer.java

```diff
@@ -76,47 +96,62 @@
-                cfg.getString("portal.dbSslKeyStorePwd"),
-                cfg.getString("portal.dbSslTrustStore"),
-                cfg.getString("portal.dbSslTrustStorePwd"));
-        DSMServer server = new DSMServer();
-
-        File test = new File(cfg.getString("portal.dbSslTrustStore"));
-        if (test.exists()) {
-            logger.info("TrustStore does exist");
+        // immediately lock isReady so that ah/start route will wait
+        synchronized (isReady) {
+            //config without secrets
+            Config cfg = ConfigFactory.load();
+            //secrets from vault in a config file
+            File vaultConfigInCwd = new File(VAULT_DOT_CONF);
+            File vaultConfigInDeployDir = new File(GAE_DEPLOY_DIR, VAULT_DOT_CONF);
+            File vaultConfig = vaultConfigInCwd.exists()  ? vaultConfigInCwd : vaultConfigInDeployDir;
+            logger.info("Reading config values from "+vaultConfig.getAbsolutePath());
+            cfg = cfg.withFallback(ConfigFactory.parseFile(vaultConfig));
+
+            if (StringUtils.isNotBlank(cfg.getString("portal.googleProjectCredentials"))) {
+                System.setProperty("GOOGLE_APPLICATION_CREDENTIALS", cfg.getString("portal.googleProjectCredentials"));
+            }
+            DSMServer server = new DSMServer();
             server.configureServer(cfg);
-
-            //            server.setupCustomDB(cfg);
-
-            logger.info("Server configuration complete.");
-        }
-        else {
-            logger.error("TrustStore does not exist");
+            isReady.set(true);
         }
     }
 
     protected void configureServer(@NonNull Config config) {
         logger.info("Property source: " + config.getString("portal.environment"));
         logger.info("Configuring the server...");
         threadPool(-1, -1, 30000);
-        logger.info("On port " + config.getInt("portal.port"));
-        port(config.getInt("portal.port"));
+        int port = config.getInt("portal.port");
+        String appEnginePort = System.getenv("PORT");
+
+        // if port is passed via env var, assume it's GAE and prefer this port
+        if (appEnginePort != null) {
+            port = Integer.parseInt(appEnginePort);
+        }
+        long bootTimeoutSeconds = DEFAULT_BOOT_WAIT.getSeconds();
+        if (config.hasPath(ApplicationConfigConstants.BOOT_TIMEOUT)) {
+            bootTimeoutSeconds = config.getInt(ApplicationConfigConstants.BOOT_TIMEOUT);
+        }
+
+
+        logger.info("Using port {}", port);
+        port(port);
+        registerAppEngineStartupCallback(bootTimeoutSeconds);
         setupDB(config);
         setupRouting(config);
+        String preferredSourceIPHeader = null;
+        if (config.hasPath(ApplicationConfigConstants.PREFERRED_SOURCE_IP_HEADER)) {
+            preferredSourceIPHeader = config.getString(ApplicationConfigConstants.PREFERRED_SOURCE_IP_HEADER);
+        }
+        JettyConfig.setupJetty(preferredSourceIPHeader);
```

Comment:
Does this `setupJetty()` call need to go before `registerAppEngineStartupCallback()`? Mounting any routes or filters will have initialized the Spark server, which is what is being done in `registerAppEngineStartupCallback()` and `setupRouting()`?

Reply:
Looking at the  DSM logs (with the real IP), it doesn't look like the order matters here.  This  order of calls still produces  the correct IP address in DSM logs.

Reply:
I still think it's a bit risky. From my reading of Spark's internals, the embedded server is initialized in a separate thread, so I suspect we might just be getting lucky that our custom factory is getting added before the thread goes through it's paces.